### PR TITLE
adds delete user report API endpoint

### PR DESCRIPTION
**Pull request by esotrope**

### CODE REVIEW

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/configuration/RemoveStaleReportsConfiguration.java [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import org.opentestsystem.rdw.reporting.processor.service.DefaultReportRemovalService;
 import org.opentestsystem.rdw.reporting.processor.service.ReportContentService;
 import org.opentestsystem.rdw.reporting.processor.service.ReportRemovalService;
-import org.opentestsystem.rdw.reporting.processor.service.ReportService;
+import org.opentestsystem.rdw.reporting.processor.service.UserReportService;
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/configuration/RemoveStaleReportsConfiguration.java]: renamed repo, service and controller to match the resource/entity name

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/UserReportController.java [hunk 2]
```diff
@@ -61,4 +64,13 @@ public ReportResource createReport(
...
     }
 
+    @DeleteMapping("{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void delete(
+            @AuthenticationPrincipal final User user,
+            @PathVariable final long id) {
+
+        service.delete(user, id);
+    }
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/UserReportController.java]: new controller method

File: report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/multithreaded/JdbcUserReportRepositoryMT.java [hunk 3]
```diff
@@ -18,35 +18,42 @@
...
+                b.copy()
                         .metadata("new value", "0")
                         .metadata("completed_chunk_count", "0")
-                        .build());
-        repository.completeChunk(report);
+                        .build()
+        );
+
+        repository.completeChunk(c);
+        repository.deleteByUserAndId("test", a.getId());
```
Comment (esotrope) [inline report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/multithreaded/JdbcUserReportRepositoryMT.java]: added method to MT

File: report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/web/UserReportControllerIT.java [hunk 4]
```diff
@@ -226,4 +227,26 @@ public void createAggregateReportShouldReturnHandle() throws Exception {
...
+
+        doThrow(new NoSuchElementException())
+                .when(service)
+                .delete(user, 1L);
+
+        mvc.perform(delete("/reports/1")
+                .contentType(MediaType.APPLICATION_JSON)
+                .with(user(user)))
+                .andExpect(status().isNotFound());
+    }
```
Comment (esotrope) [inline report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/web/UserReportControllerIT.java]: new endpoint tests

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcUserReportRepository.java [hunk 5]
```diff
@@ -167,7 +168,23 @@ public UserReport findById(final long reportId) {
...
+    public void deleteByUserAndId(String user, long id) {
+        final int updatedCount = template.update(
+                deleteByUserAndId,
+                new MapSqlParameterSource()
+                .addValue("user_login", user)
+                .addValue("id", id)
+        );
+        if (updatedCount != 1) {
+            throw new NoSuchElementException("Unable to delete report");
+        }
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcUserReportRepository.java]: new repo method

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java [hunk 6]
```diff
@@ -67,4 +67,13 @@
      * @param ids The ids of the reports we wish to delete
      */
     void deleteIds(Collection<Long> ids);
+
+    /**
+     * Deletes reports with the provied user and ID
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java]: nit: typo

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java [hunk 7]
```diff
@@ -67,4 +67,13 @@
...
+     * Throws an error if no record is found matching this user/ID combination
+     *
+     * @param user The user
+     * @param id The user report ID
+     */
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java]: nit: could put a `@throws NoSuchElementException`

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/DefaultExamReportService.java [hunk 8]
```diff
@@ -77,13 +77,19 @@ public ReportContentResource openReportContent(@NotNull final User user, final l
...
+    @Override
     public UserReport createReport(
             @NotNull final User user,
             @NotNull final ReportQuery query) {
         checkPermissions(user.getPermissionsById(), query.getType());
         return generator.generateReport(query, user);
     }
 
+    @Override
+    public void delete(User user, long id) {
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/DefaultExamReportService.java]: nit: `final`

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java [hunk 9]
```diff
@@ -47,4 +47,12 @@
      */
     UserReport createReport(@NotNull User user, @NotNull ReportQuery query);
 
+    /**
+     * Deletes a report matching the given user and user report ID
+     *
+     * @param user The user
+     * @param id The user report ID
+     */
```
Comment (malaffoon) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: nit: `@throws` ?

Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: Or should the service layer eat the NoSuchElementException and just silently no-op? Maybe log.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java (hunk 7 above)
Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java]: hm.. i recall you recommending in the past to not add `@throws` for unchecked exceptions

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java (hunk 9 above)
Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: depends, do we want to send hackers 404s or 200s that affect nothing?

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java (hunk 7 above)
Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java]: I'm fine and good with the `@throws` in the javadocs; what i don't think we need is the `throws` definition in the java code for unchecked exceptions.

**Comment by esotrope** [comment]
> A thought for future PR's like this (and feel free to parrot this back to me when i do it): when doing a rename, first make a PR that has _just_ the rename in it, and tell reviewers that it is just a rename and requires a cursory review. Then make the real changes with a much smaller affected file set.

yeah agreed i'll try and cool my engines a bit and send things in better chunks 👍

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java (hunk 7 above)
Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/UserReportRepository.java]: ah that sounds good. i'm on board

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java (hunk 9 above)
Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: Hmm. I think 200's. The 404 lets them probe the system. I guess we could do something like:
```
if delete fails to affect just 1 row:
    find the user report; if the find fails:
        log.debug("attempt to delete unknown user report...")
        return quietly
    else (found report):
        log.warn("failed to delete user report...")
        throw OMG
```
Just thinking out loud.

Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: cool if i just make a jira to follow up on this for now?

Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/service/ExamReportService.java]: Of course.

**Pull request merged**